api/types.go

```diff
@@ -141,20 +143,49 @@ type SeriesType string
 // end =   0 mod resolution
 //
 // This range is inclusive of Start and End (i.e. [Start, End]). Start and End
-// are Unix second timestamps. Resolution is in seconds.
-// TODO: Make these ms
+// are Unix milliseconds timestamps. Resolution is in milliseconds.
+// (Millisecond precision allows an effective range of 290 million years in each direction)
 type Timerange struct {
-	Start      int64
-	End        int64
-	Resolution int64
+	start      int64
+	end        int64
+	resolution int64
 }
 
-// IsValid determines whether the given timerange meets the constraint.
-func (tr Timerange) IsValid() bool {
-	return tr.Resolution > 0 &&
-		tr.Start%tr.Resolution == 0 &&
-		tr.End%tr.Resolution == 0 &&
-		tr.Start <= tr.End
+// DefaultTimerange is a valid timerange which can be safely used
+func DefaultTimerange() Timerange {
+	return Timerange{start: 0, end: 0, resolution: 30}
+}
+
+// Start() returns the .start field
+func (tr Timerange) Start() int64 {
+	return tr.start
+}
+
+// End() returns the .end field
+func (tr Timerange) End() int64 {
+	return tr.end
+}
+
+// Resolution() returns the .resolution field
+func (tr Timerange) Resolution() int64 {
+	return tr.resolution
+}
+
+// NewTimerange creates a timerange which is validated, providing error otherwise.
+func NewTimerange(start, end, resolution int64) (*Timerange, error) {
```

Comment:
I think it's unidiomatic go to give pointers to simple structs - probably less efficient too.

that said, the topic of whether we should return pointers or not in many different places is a highly debatable topic - i'm still not sure which case should be which.